File: cpp/submod/SetCover.cpp
#include<iostream>
#include<vector>
#include<string>
#include<algorithm>
#include<cmath>
#include<set>
#include<iterator>
#include<map>
#include"SetCover.h"

SetCover::SetCover(ll n_, std::vector<std::unordered_set<int>> const &coverSet_, int numConcepts_, std::vector<float> const& conceptWeights_): n(n_), coverSet(coverSet_), numConcepts(numConcepts_), conceptWeights(conceptWeights_)  {
}

// SetCover* SetCover::clone() {
// 	return NULL;
// }

double SetCover::evaluate(std::unordered_set<ll> const &X) {
	double result=0;

	if(X.size()==0) {
		return 0;
	}

	std::unordered_set<int> conceptsCovered;
	for(auto elem: X) {
		for(auto con: coverSet[elem]) {
			conceptsCovered.insert(con);
		}
	}
	for(auto con: conceptsCovered) {
		result += conceptWeights[con];
	}

	return result;
}

double SetCover::evaluateWithMemoization(std::unordered_set<ll> const &X) { 
	double result = 0;
	if(X.size()==0) {
		return 0;
	}
	
	for(auto con: conceptsCoveredByX) {
		result += conceptWeights[con];
	}
	return result;
}


double SetCover::marginalGain(std::unordered_set<ll> const &X, ll item) {
	double gain = 0;
	if (X.find(item)!=X.end()) {
		return 0;
	}
    std::unordered_set<int> conceptsCovered;
	for(auto elem: X) {
		for(auto con: coverSet[elem]) {
			conceptsCovered.insert(con);
		}
	}
	for(auto con: coverSet[item]) {
        if(conceptsCovered.find(con) == conceptsCovered.end()) {
            gain += conceptWeights[con];
		}
	}
	return gain;
}


double SetCover::marginalGainWithMemoization(std::unordered_set<ll> const &X, ll item, bool enableChecks) {
	double gain = 0;
	if (enableChecks && X.find(item)!=X.end()) {
		return 0;
	}
	//std::cout << "Concepts covered by " << item << " = {";
	for(auto con: coverSet[item]) {
		//std::cout << concept << ", ";
        if(conceptsCoveredByX.find(con) == conceptsCoveredByX.end()) {
            gain += conceptWeights[con];
		}
	}
	//std::cout << "}\n";
	//std::cout << "Gain = " << gain << "\n";
	return gain;
}

void SetCover::updateMemoization(std::unordered_set<ll> const &X, ll item) {
	if (X.find(item)!=X.end()) {
		return;
	}
	for(auto con: coverSet[item]) {
		conceptsCoveredByX.insert(con);
	}
}

std::unordered_set<ll> SetCover::getEffectiveGroundSet() {
	std::unordered_set<ll> effectiveGroundSet;
	effectiveGroundSet.reserve(n);
	for (ll i = 0; i < n; ++i){
		effectiveGroundSet.insert(i); 
	}
	return effectiveGroundSet;
}


// std::vector<std::pair<ll, double>> SetCover::maximize(std::string optimizer,ll budget, bool stopIfZeroGain=false, bool stopIfNegativeGain=false, float epsilon = 0.1, bool verbose=false, bool showProgress=true) {
// 	if(optimizer == "NaiveGreedy") {
// 		return NaiveGreedyOptimizer().maximize(*this, budget, stopIfZeroGain, stopIfNegativeGain, verbose, showProgress);
// 	} else if(optimizer == "LazyGreedy") {
//         return LazyGreedyOptimizer().maximize(*this, budget, stopIfZeroGain, stopIfNegativeGain, verbose, showProgress);
// 	} else if(optimizer == "StochasticGreedy") {
//         return StochasticGreedyOptimizer().maximize(*this, budget, stopIfZeroGain, stopIfNegativeGain, epsilon, verbose, showProgress);
// 	} else if(optimizer == "LazierThanLazyGreedy") {
//         return LazierThanLazyGreedyOptimizer().maximize(*this, budget, stopIfZeroGain, stopIfNegativeGain, epsilon, verbose, showProgress);
// 	} else {
// 		throw "Error: Invalid Optimizer";
// 	}
// }

void SetCover::clearMemoization() {
	conceptsCoveredByX.clear();
}

void SetCover::setMemoization(std::unordered_set<ll> const &X) 
{
    clearMemoization();
    std::unordered_set<ll> temp;
    for (auto elem: X)
	{	
		updateMemoization(temp, elem);
		temp.insert(elem);	
	}
}

File: pytorch/submod/SetCover.py
import torch
import torch.nn as nn
import numpy as np
import random
from ..SetFunction import SetFunction

class SetCoverFunction(SetFunction):
    def __init__(self, n, cover_set, num_concepts, concept_weights = None):
        super(SetFunction, self).__init__()
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        self.n = n
        self.cover_set = cover_set
        self.num_concepts = num_concepts
        self.concept_weights = concept_weights
        if self.concept_weights is None:
            self.concept_weights = [1.0] * num_concepts
        else:
            self.concept_weights = torch.tensor(concept_weights, dtype=torch.float32).to(device)


        self.concepts_covered_by_x = set()


    def evaluate(self, X):
      result = 0.0

      if X.numel() == 0:
          return 0.0

      concepts_covered = set()
      for elem in X:
          concepts_covered.update(self.cover_set[elem.item()])

      for con in concepts_covered:
          result += self.concept_weights[con]

      return result


    def evaluate_with_memoization(self, X):
        result = 0.0

        if X.numel() == 0:
            print("hi")
            return 0.0

        for con in self.concepts_covered_by_x:
            result += self.concept_weights[con]
            print(result)

        return result

    def marginal_gain(self, X, item):
        gain = 0.0

        if item in X:
            return 0.0

        concepts_covered = set()
        for elem in X:
            concepts_covered.update(self.cover_set[elem])

        for con in self.cover_set[item]:
            if con not in concepts_covered:
                gain += self.concept_weights[con]

        return gain.item()

    def marginal_gain_with_memoization(self, X, item, enable_checks=True):
        gain = 0.0

        if enable_checks and item in X:
            return 0.0
        for con in self.cover_set[item]:
          if con not in self.concepts_covered_by_x:
                gain += self.concept_weights[con]

        return gain

    def update_memoization(self, X, item):
        if item in X:
            return

        self.concepts_covered_by_x.update(self.cover_set[item])

    def get_effective_ground_set(self):
        return set(range(self.n))

    def clear_memoization(self):
        self.concepts_covered_by_x.clear()

    def set_memoization(self, X):
        self.clear_memoization()
        temp = set()
        for elem in X:
            self.update_memoization(temp, elem)
            temp.add(elem)
